Added support for CommonJS require()

haberman opened the pull request:
This required several codegen changes and a bunch of glue in gulp.js to test it all.

joeltine commented on js/README.md:
```diff
@@ -1,14 +1,134 @@
+2. Closure-style imports (eg. `goog.require('my.package.MyProto');`)
+
+Support for ES6-style imports is not implemented yet.  Browsers can
+be supported by using Browserify, webpack, Closure Compiler, etc. to
```
Add a compiler link? https://github.com/google/closure-compiler/wiki/JS-Modules#es6-modules

haberman replied:
Hmm, I don't think the Closure Compiler-specific support for ES6 is the relevant information here. When we add ES6 support, it will be so people can use any ES6-native tooling (Babel, Closure Compiler, VMs that support ES6 directly, etc). Closure Compiler is only one such example.

joeltine commented on js/README.md:
```diff
@@ -1,14 +1,134 @@
+   `npm install google-protobuf`, or use the files in this directory.
+2. The Protocol Compiler `protoc`.  This translates `.proto` files
+   into `.js` files.  The compiler is not currently available via
+   npm -- you must download and compile it from GitHub or a tarball.
```
A link to the protoc downloads would be nice.

haberman replied:
Done.

joeltine commented on js/README.md:
```diff
@@ -1,14 +1,134 @@
+If you want to use Closure imports, your build should run a command
+like this:
+
+    $ protoc --js_out=library=myproto_libs,binary:. messages.proto base.proto
```
I'm confused, I downloaded the pre-built protoc binary from https://github.com/google/protobuf/releases and I don't see `--js_out` when I run `protoc --help`.

joeltine replied:
It's worth mentioning, this came from advice in their docs: 

> For non-C++ users, the simplest way to install the protocol compiler is to download a pre-built binary from our release page:

joeltine replied:
Maybe it's my lack of experience with protoc, but the `js_out` flag seems very unintuitive to me.  Hopefully this is documented somewhere and can be linked here.

haberman replied:
Thanks for the note about the pre-built binaries -- I didn't actually know about these! I agree those are much more convenient than building your own.

Are you sure the most recent protoc doesn't include `--js_out` in `--help`? I just tried downloading the latest binary and I saw it there.

Sure, I can add some words about how the js_out flag works. I agree it's not super intuitive.

joeltine replied:
> Sure, I can add some words about how the js_out flag works. I agree it's not super intuitive.

I don't see anything on `js_out`.

joeltine commented on js/commonjs/export_asserts.js:
```diff
@@ -0,0 +1,33 @@
+  }
+}
+
+exports.COMPILED = COMPILED
```
Maybe add mention where `COMPILED` comes from.

haberman replied:
Done.

joeltine commented on js/commonjs/export.js:
```diff
@@ -0,0 +1,16 @@
+// These are used by generated code but should not be used directly by clients.
+exports.exportSymbol = goog.exportSymbol;
+exports.inherits = goog.inherits;
+exports.object = {extend: goog.object.extend};
```
Maybe it's my lack of Node experience, but don't you need to require what you use here (e.g., `goog.object` etc.)?

haberman replied:
Good point. Done, except for a few things (`goog.inherits`, `goog.exportSymbol`) which are in Closure base and can't be `goog.require()`'d.

joeltine replied:
Missing `require('google-closure-library')`?

joeltine replied:
... or that's brought in via the Jasmine "helpers"?

haberman replied:
I don't think we want `require('google-closure-library')` -- CommonJS users won't have it, because the parts of Closure we need will be "statically linked" into `google-protobuf.js`.

joeltine commented on js/commonjs/export_asserts.js:
```diff
@@ -0,0 +1,33 @@
+for (var key in global) {
+  if ((typeof key == "string") && global.hasOwnProperty(key) &&
+      shouldExport(key)) {
+    exports[key] = global[key];
```
Ugh. I wish this weren't necessary. The fact the assertions are global is ugly. Honestly what we should do is put all the assert methods on `goog.testing.asserts`.

haberman replied:
Yes agreed -- get it changed in Closure and I'll be happy to change it here. :)

joeltine commented on js/commonjs/rewrite_tests_for_commonjs.js:
```diff
@@ -0,0 +1,69 @@
+  var is_loadfromfile = line.match(/CommonJS-LoadFromFile: ([^ ]*) (.*)/);
+  var is_settestonly = line.match(/goog.setTestOnly()/);
+  if (is_settestonly) {
+    // Remove this line.
```
?

haberman replied:
Lines in the input that say `goog.setTestOnly()` are removed from the output completely since that function isn't available and isn't necessary to call.

joeltine replied:
I guess it's a nit, but the last else could just be `else if(!settestonly)`. Empty if's always feel weird to me.

haberman replied:
Done.

joeltine commented on js/commonjs/rewrite_tests_for_commonjs.js:
```diff
@@ -0,0 +1,69 @@
+      console.log("var global = Function('return this')();");
+      console.log("");
+      console.log("// Bring asserts into the global namespace.");
+      console.log("google_protobuf.object.extend(global, asserts);");
```
Why are these `google_protobuf` exports needed? Can't you just use `google-closure-library` directly? In other words:

```
require("google-closure-library");
goog.require('goog.object');
...
```

joeltine replied:
I guess, I'm a little confused on the design. If you're bringing the asserts into the global namespace anyway, why do you need the shim in `export_asserts.js`?

haberman replied:
The overall goal of this change is to provide a simple out-of-the-box experience for people who are not using Closure Compiler or Closure-style require/provide.

Leaving `goog.require()` in the source forces users to generate `deps.js` and make sure all the paths line up. If `goog.require()` fails, they have to dig into this unfamiliar system and figure out why. I went through the pain of this when setting up our Closure-based tests, and had to write this somewhat hacky script to help Closure find files in the current directory: https://github.com/google/protobuf/blob/master/js/node_loader.js. (This was when I was already using `goog/bootstrap/nodejs.js`, see: https://github.com/google/protobuf/blob/master/js/jasmine.json#L8).

With the approach in this CL, the entire protobuf runtime library gets compiled to a single `google-protobuf.js` that does not depend on any `deps.js` or external libraries. This also makes things more convenient for browsers: they just need one `<script>` tag for the protobuf runtime library.

The future is ES6 imports anyway, which are file-oriented. I think moving away from symbol-based includes and towards file-based includes like `require` is moving things in the right direction for convenience and for the future. At this point I see `goog.require()` and `deps.js` as a legacy setup, and I don't want to push a legacy setup on our users.

haberman replied:
`export_asserts.js` takes symbols that are global inside that module and exports them. That allows a test to say:

```javascript
var asserts = require('closure_asserts_commonjs');

// Now this symbol is available:
asserts.assertObjectEquals(1, 1);
```

But the tests do not call the asserts like `asserts.assertObjectEquals()`, they just say `assertObjectEquals()`. So we need to take the keys in the `asserts` object and put them in the global namespace of the test.

Put another way, both are necessary because the global namespace of the module is not the same as the global namespace of the test. The module can't put things in the test's global namespace. CommonJS modules are more hygienic than Closure modules in this way.

joeltine commented on js/commonjs/rewrite_tests_for_commonjs.js:
```diff
@@ -0,0 +1,69 @@
+    }
+  } else if (is_loadfromfile) {
+    if (!module) {
+      console.log("var google_protobuf = require('google-protobuf');");
```
I would think camel case would be the typical Node/JS style for var names.

haberman replied:
Done (changed to `googleProtobuf`).

joeltine commented on js/commonjs/rewrite_tests_for_commonjs.js:
```diff
@@ -0,0 +1,69 @@
+/**
+ * @fileoverview Utility to translate test files to CommonJS imports.
```
I'm wondering if this translation step is needed at all. Closure already has a [bootstrap script](https://github.com/google/closure-library/blob/master/closure/goog/bootstrap/nodejs.js) that effectively converts `goog.require` to CommonJS `require`. All you need is a deps.js file to do the translation from namspace to src file. The bootstrap script linked tells the whole story and gives examples.

haberman replied:
The existing tests use this approach, and this PR is not removing the existing tests. This PR is adding new functionality and new tests for native CommonJS functionality that doesn't depend on `goog.require()` or `deps.js`.

haberman commented:
I added a commit to address your comments. PTAL.

joeltine commented on js/commonjs/rewrite_tests_for_commonjs.js:
```diff
@@ -0,0 +1,69 @@
+var pkg = null;
+lineReader.on('line', function(line) {
+  var is_require = line.match(/goog\.require\('([^']*)'\)/);
+  var is_loadfromfile = line.match(/CommonJS-LoadFromFile: ([^ ]*) (.*)/);
```
Isn't `[^ ]` basically `\S`?

joeltine replied:
In general, your JS is a big mix of camel case and underscore naming. Be consistent if you can.

haberman replied:
Done. Good catch on that -- I don't write enough JavaScript regularly to be used to camelCasing. :)

joeltine commented on js/commonjs/rewrite_tests_for_commonjs.js:
```diff
@@ -0,0 +1,69 @@
+ *
+ * This script parses that special comment and uses it to generate proper
+ * CommonJS require() statements so that the tests can run and pass using
+ * CommonJS imports.
```
An example in the comment of how the the above `goog.require`s are translated to CommonJS `require` would be helpful.

haberman replied:
Done.

joeltine commented on js/README.md:
```diff
@@ -1,14 +1,136 @@
+You can compile `protoc` from GitHub or a source tarball.  From the
+top level directory type:
+
+    $ ./autogen.sh  (only necessary for GitHub)
```
I cloned a copy of this PR and tried to follow these instructions. `autogen.sh` failed because I was missing some build dependencies only discussed in the C++ Installation page: https://github.com/google/protobuf/blob/master/src/README.md. Maybe add a link?

haberman replied:
Good point -- since binary downloads are available too, and it's easier, I now recommend that and just link to the top-level README for users who want to compile from source.

joeltine commented on js/README.md:
```diff
@@ -1,14 +1,136 @@
+
+Once you have `protoc` compiled, you can run the tests by typing:
+
+    $ npm install
```
So at this point in the instructions, I'm in the base directory after building protoc,but `npm install` only works in `js/`. Would help to add a mention to `cd js/` somewhere.

haberman replied:
Done.

joeltine commented on js/binary/reader_test.js:
```diff
@@ -42,6 +42,8 @@
 
 goog.require('goog.testing.asserts');
+
+// CommonJS-LoadFromFile: google_protobuf
```
`googleProtobuf`... I ran the tests with this file as-is, and everything passed... how is that possible?

haberman replied:
Ah, it's because for CommonJS I don't currently test 100% of the `*_test.js` files. Several of the `binary/*_test.js` files access internal symbols from the library (or extra symbols from Google Closure) and making all of that work with CommonJS seemed to cross a line of being more work than it was worth.

But it's a good catch -- if we're not testing `reader_test.js`, there is no point in adding this line to it. So I removed this line from `reader_test.js` and `utils_test.js`.

I tried to add a comment to `commonjs/jasmine.json` to make it more clear that we're leaving out some files. But unfortunately it is JSON and cannot have comments.

haberman commented:
I think I addressed all comments. I also made the translation script properly camelCase. PTAL, thanks!

joeltine commented on js/README.md:
```diff
@@ -1,14 +1,135 @@
+follow the instructions in [the top-level
+README](https://github.com/google/protobuf/blob/master/src/README.md).
+
+Once you have `protoc` compiled, you can run the tests by typing:
```
Is it worth mentioning `protoc` needs to be on the $PATH?

joeltine commented:
One small remaining nit comment, but LGTM overall!

haberman commented:
I added more info about where `npm test` will look for `protoc` (and how to override that). I also added docs about `--js_out` (forgot to do this before).

Want to take one more look before I merge? Thanks!
